arch/sim/src/sim/up_i2cbuslinux.c

```diff
@@ -0,0 +1,288 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <sys/types.h>
+#include <sys/ioctl.h>
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <syslog.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <linux/i2c-dev.h>
+#include <linux/i2c.h>
+
+#include "up_i2cbus.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define ERROR(fmt, ...) \
+        syslog(LOG_ERR, "up_i2cbuslinux: " fmt "\n", ##__VA_ARGS__)
+#define INFO(fmt, ...) \
+        syslog(LOG_ERR, "up_i2cbuslinux: " fmt "\n", ##__VA_ARGS__)
+#define DEBUG(fmt, ...)
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct linux_i2cbus_master_s
+{
+  const struct i2c_ops_s *ops; /* I2C vtable */
+  int file;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+int linux_i2cbus_transfer_host(struct i2c_master_s *dev,
+                               struct i2c_msg_s *msgs, int count);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+struct i2c_ops_s i2c_linux_ops =
+{
+  .transfer = linux_i2cbus_transfer_host,
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: linux_i2cbus_transfer_host
+ *
+ * Description:
+ *   Provide i2c transfer
+ *
+ ****************************************************************************/
+
+int linux_i2cbus_transfer_host(struct i2c_master_s *dev,
+                               struct i2c_msg_s *msgs, int count)
+{
+  int ret = 0;
+  struct linux_i2cbus_master_s *priv = (struct linux_i2cbus_master_s *)dev;
+  int file = priv->file;
+  uint8_t *pack_buf = NULL;
+  struct i2c_rdwr_ioctl_data ioctl_data;
+  struct i2c_msg l_msgs[2];  /* We only support up to 2 messages */
```

Comment:
there are one to one mapping between Linux and NuttX i2c msg defintion, why not fully convert like this?
```
struct i2c_msg l_msgs[count]; 
for (i = 0; i < count; i++)
{
  l_msgs[i].addr = msgs[i].addr;
  l_msgs[i].buf = msgs[i].buffer;
  l_msgs[i].len = msgs[i].length;
  l_msgs[i].flags = 0;
  if (msgs[i].flags & NUTTX_I2C_M_READ)
    {
      l_msgs[i].flags |= I2C_M_RD;
    }
  if (msgs[i].flags & NUTTX_I2C_M_TEN)
    {
      l_msgs[i].flags |= I2C_M_TEN;
    }
  if (msgs[i].flags & NUTTX_I2C_M_NOSTART)
    {
      l_msgs[i].flags |= I2C_M_NOSTART;
    }
  if (!(msgs[i].flags & I2C_M_NOSTOP))
    {
      l_msgs[i].flags |= I2C_M_STOP;
    }
}
```

Reply:
There are absolutely not 1-1 mapping see all the logic that I have to do in this function, it is not trivial.

Getting this right is what took all the time.

Reply:
for instance I2C_M_NOSTART is not a thing on Linux you need to make a real transfer.  We also have to change the READ and WRITE to align with what is expected.

Reply:
Yes, but we should support as much as possible(e.g. message count and avoid buffer copy)

Reply:
We have to do the packing there is no way around it without getting extra START and STOP